libraries/AP_Quicktune/AP_Quicktune.cpp

```diff
@@ -0,0 +1,686 @@
+        // Update slew gain
+        if (slew_parm != param_s::END){
+            float P = get_param_value(slew_parm);
+            axis_names axis = get_axis(slew_parm);
+            // local ax_stage = string.sub(slew_parm, -1)
+            adjust_gain(slew_parm, P+slew_delta);
+            slew_steps = slew_steps - 1;
+            write_quik(get_slew_rate(axis), P, slew_parm);
+            if (slew_steps == 0){
+                GCS_SEND_TEXT(MAV_SEVERITY_INFO, "%s %.4f", get_param_name(slew_parm), P);
+                slew_parm = param_s::END;
+                if (get_current_axis() == axis_names::DONE){
+                    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Tuning: DONE");
+                    tune_done_time = now;
+                }
+            }
+        }
+        return;
+    }
+
+    axis_names axis = get_current_axis();
+
+    if (axis == axis_names::DONE){
+        // Nothing left to do, check autosave time
+        if (tune_done_time != 0 && auto_save > 0){
+            if (now - tune_done_time > (auto_save*1000)){
+                need_restore = false;
+                save_all_params();
+                GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Tuning: Saved");
+                tune_done_time = 0;
+            }
+        }
+        return;
+    }
+
+    if (!need_restore){
+        need_restore = true;
+        // We are just starting tuning, get current values
+        GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Tuning: Starting tune");
+        // Get all params
+        for (int8_t pname = 0; pname < uint8_t(param_s::END); pname++){
+            param_saved[pname] = get_param_value(param_s(pname));
+        }
+        // Set up SMAX
+        param_s is[3];
+        is[0] = param_s::RLL_SMAX;
+        is[1] = param_s::PIT_SMAX;
+        is[2] = param_s::YAW_SMAX;
+        for (uint8_t i = 0; i < 3; i++){
+            float smax = get_param_value(is[i]);
+            if (smax <= 0){
+                adjust_gain(is[i], DEFAULT_SMAX); 
+            }
+        }
+    }
+
+    if (now - last_pilot_input < PILOT_INPUT_DELAY){
+        return;
+    }
+
+    if (!item_in_bitmask(uint8_t(axis), filters_done)){
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "Starting %s tune", get_axis_name(axis));
+        setup_filters(axis);
+    }
+
+    param_s pname = get_pname(axis, current_stage);
+    float P = get_param_value(pname);
+    float limit = gain_limit(pname);
+    bool limited = (limit > 0.0 && P >= limit);
+    float srate = get_slew_rate(axis);
+    bool oscillating = srate > osc_smax;
+    
+    // Check if reached limit
+    if (limited || oscillating){
+        float reduction = (100.0-gain_margin)*0.01;
+        if (!oscillating){
+            reduction = 1.0;
+        }
+        float new_gain = P * reduction;
+        if (limit > 0.0 && new_gain > limit){
+            new_gain = limit;
+        }
+        float old_gain = param_saved[uint8_t(pname)];
+        if (new_gain < old_gain && (pname == param_s::PIT_D || pname == param_s::RLL_D)){
```

Comment:
The algorithm here needs to be documented somewhere, I'm not sure I fully understand what it is doing and it makes it hard to scrutinze its effectiveness

Reply:
comments added